precompiles/Modexp.yul

```diff
@@ -333,6 +333,54 @@ object "ModExp" {
 
             }
 
+            function getLimbValueAtOffset(limbPointer, anOffset) -> limbValue {
+                limbValue := mload(add(anOffset, limbPointer))
+            }
+
+            function storeLimbValueAtOffset(limbPointer, anOffset, aValue) {
+                    mstore(add(limbPointer, anOffset), aValue)
+            }
+
+            /// @notice Computes the difference between two 256 bit number and keeps
+            /// account of the borrow bit
+            /// in lshPointer and rhsPointer.
+            /// @dev Reference: https://github.com/lambdaclass/lambdaworks/blob/main/math/src/unsigned_integer/element.rs#L785
+            /// @param leftLimb The left side of the difference (i.e. the a in a - b).
+            /// @param rightLimb The right side of the difference (i.e. the b in a - b).
+            /// @return subtractionResult i.e. the c in c = a - b.
+            /// @return returnBorrow If there was any borrow on the subtraction, is returned as 1.
+            function subLimbsWithBorrow(leftLimb, rightLimb, limbBorrow) -> subtractionResult, returnBorrow {
+                let rightPlusBorrow := add(rightLimb, limbBorrow)
+                subtractionResult := sub(leftLimb, rightPlusBorrow)
+                if gt(subtractionResult, leftLimb) {
+                 returnBorrow := 1
+                }
+            }
+            /// @notice Computes the BigUint subtraction between the number stored
+            /// in lshPointer and rhsPointer.
+            /// @dev Reference: https://github.com/lambdaclass/lambdaworks/blob/main/math/src/unsigned_integer/element.rs#L795
+            /// @param lhsPointer The start of the left hand side subtraction Big Number.
+            /// @param rhsPointer The start of the right hand side subtraction Big Number.
+            /// @return numberOfLimbs The number of limbs of both numbers.
+            /// @return resultPointer Where the result will be stored.
+            function bigUintSubtractionWithBorrow(lhsPointer, rhsPointer, numberOfLimbs, resultPointer) -> resultPointer, borrow {
+                let leftIthLimbValue
+                let rightIthLimbValue
+                let ithLimbBorrowResult
+                let ithLimbSubtractionResult
+                let borrow := 0
+                let limbOffset := 0
+                for {let i := numberOfLimbs} gt(i, 0) {i := sub(i, 1)} {
+                    limbOffset := mul(sub(i,1), 32)
+                    leftIthLimbValue := getLimbValueAtOffset(lhsPointer, limbOffset)
+                    rightIthLimbValue := getLimbValueAtOffset(rhsPointer, limbOffset)
+                    ithLimbSubtractionResult, borrow :=
+                                               subLimbsWithBorrow(leftIthLimbValue, rightIthLimbValue, borrow)
+                    storeLimbValueAtOffset(resultPointer, limbOffset, ithLimbSubtractionResult)
+
+                }
```

Comment:
We can get rid of `numberOfLimbs - 1` `mul` operations if we start `limbOffset` at `shl(5, numberOfLimbs)` and adjust it subtracting LIMB_SIZE_IN_BYTES() at the end of every iteration.
```suggestion
                let limbOffset := shl(5, numberOfLimbs)
                for {let i := numberOfLimbs} gt(i, 0) {i := sub(i, 1)} {
                    leftIthLimbValue := getLimbValueAtOffset(lhsPointer, limbOffset)
                    rightIthLimbValue := getLimbValueAtOffset(rhsPointer, limbOffset)
                    ithLimbSubtractionResult, borrow :=
                                               subLimbsWithBorrow(leftIthLimbValue, rightIthLimbValue, borrow)
                    storeLimbValueAtOffset(resultPointer, limbOffset, ithLimbSubtractionResult)
                    limbOffset := sub(limbOffset, LIMB_SIZE_IN_BYTES())
                }
```

Reply:
Sure, I'll test this again after the suggested changes and let you know.

Reply:
These changes make the algorithm give the correct output, I'll ignore them for now. Are they really needed? @ilitteri